File: backend/panel_auth.py
"""Autentificare panou: utilizator, sesiune, 2FA (TOTP / Telegram)."""

import os
import random
import secrets
from datetime import timedelta

import pyotp
from werkzeug.security import check_password_hash, generate_password_hash

from models import PanelSession, PanelUser, TwoFaChallenge, utcnow

SESSION_HOURS = int(os.environ.get("PANEL_SESSION_HOURS", "24"))
CHALLENGE_MINUTES = 5
TOTP_ISSUER = os.environ.get("PANEL_TOTP_ISSUER", "NeoHost Security")


def hash_password(password):
    return generate_password_hash(str(password))


def verify_password(user, password):
    return check_password_hash(user.password_hash, str(password))


def ensure_default_admin(db):
    if db.query(PanelUser).first():
        return None
    username = os.environ.get("PANEL_ADMIN_USERNAME", "admin")
    password = os.environ.get("PANEL_ADMIN_PASSWORD", "admin")
    user = PanelUser(
        username=username,
        password_hash=hash_password(password),
        two_fa_method="none",
    )
    db.add(user)
    db.commit()
    return user


def cleanup_expired_auth(db):
    now = utcnow()
    db.query(PanelSession).filter(PanelSession.expires_at < now).delete()
    db.query(TwoFaChallenge).filter(TwoFaChallenge.expires_at < now).delete()
    db.commit()


def create_panel_session(db, user_id, ip_address="", user_agent=""):
    cleanup_expired_auth(db)
    token = secrets.token_urlsafe(32)
    expires = utcnow() + timedelta(hours=SESSION_HOURS)
    db.add(PanelSession(
        token=token,
        user_id=user_id,
        expires_at=expires,
        ip_address=(ip_address or "")[:45],
        user_agent=(user_agent or "")[:512],
    ))
    db.commit()
    return token, expires


def get_panel_session(db, token):
    if not token:
        return None
    row = db.query(PanelSession).filter_by(token=token).first()
    if not row or row.expires_at < utcnow():
        return None
    return db.query(PanelUser).filter_by(id=row.user_id, is_active=True).first()


def revoke_panel_session(db, token):
    row = db.query(PanelSession).filter_by(token=token).first()
    if row:
        db.delete(row)
        db.commit()


def list_user_sessions(db, user_id):
    cleanup_expired_auth(db)
    now = utcnow()
    rows = (
        db.query(PanelSession)
        .filter_by(user_id=user_id)
        .filter(PanelSession.expires_at >= now)
        .order_by(PanelSession.created_at.desc())
        .all()
    )
    return rows


def revoke_session_token(db, user_id, token):
    row = db.query(PanelSession).filter_by(token=token, user_id=user_id).first()
    if row:
        db.delete(row)
        db.commit()
        return True
    return False


def revoke_other_sessions(db, user_id, keep_token):
    cleanup_expired_auth(db)
    q = db.query(PanelSession).filter_by(user_id=user_id)
    if keep_token:
        q = q.filter(PanelSession.token != keep_token)
    deleted = q.delete()
    db.commit()
    return deleted


def get_primary_panel_user(db):
    return db.query(PanelUser).filter_by(is_active=True).order_by(PanelUser.id).first()


def create_2fa_challenge(db, user, method, purpose="login", code=None):
    cleanup_expired_auth(db)
    token = secrets.token_urlsafe(24)
    expires = utcnow() + timedelta(minutes=CHALLENGE_MINUTES)
    if method == "telegram" and not code:
        code = f"{random.randint(100000, 999999)}"
    db.add(TwoFaChallenge(
        token=token,
        user_id=user.id,
        code=str(code or ""),
        method=method,
        purpose=purpose,
        expires_at=expires,
    ))
    db.commit()
    return token, code, expires


def verify_2fa_challenge(db, challenge_token, user, code, method=None):
    row = db.query(TwoFaChallenge).filter_by(token=challenge_token, user_id=user.id).first()
    if not row or row.expires_at < utcnow():
        return False
    use_method = method or row.method
    if row.method != use_method:
        return False
    entered = str(code or "").strip().replace(" ", "")
    ok = False
    if use_method == "telegram":
        ok = row.code == entered
    elif use_method == "totp":
        if user.totp_secret:
            totp = pyotp.TOTP(user.totp_secret)
            ok = totp.verify(entered, valid_window=1)
    if ok:
        db.delete(row)
        db.commit()
    return ok


def generate_totp_setup(user):
    secret = pyotp.random_base32()
    totp = pyotp.TOTP(secret)
    uri = totp.provisioning_uri(name=user.username, issuer_name=TOTP_ISSUER)
    return secret, uri


def verify_totp_code(user, code):
    if not user.totp_secret:
        return False
    totp = pyotp.TOTP(user.totp_secret)
    return totp.verify(str(code or "").strip().replace(" ", ""), valid_window=1)


def login_requires_2fa(user):
    return (user.two_fa_method or "none") in ("totp", "telegram")


def session_response(user, token, expires):
    return {
        "access_token": token,
        "expires_at": expires.isoformat(),
        "user": user.to_dict(),
        "auth_type": "panel",
    }

File: hub/src/lib/panelAuth.js
import { randomBytes } from "node:crypto";
import bcrypt from "bcryptjs";
import { authenticator } from "otplib";
import { config } from "../config.js";
import { exec, queryAll, queryOne, getDb, nowIso } from "../db/index.js";

export const CHALLENGE_MINUTES = 10;

function isoAfterHours(h) {
  return new Date(Date.now() + h * 3600 * 1000).toISOString();
}

function isoAfterMinutes(m) {
  return new Date(Date.now() + m * 60 * 1000).toISOString();
}

export function hashPassword(password) {
  return bcrypt.hashSync(String(password), 10);
}

export function verifyPassword(user, password) {
  return bcrypt.compareSync(String(password), user.password_hash || "");
}

export async function ensureDefaultAdmin() {
  const first = await queryOne("SELECT id FROM panel_users LIMIT 1");
  if (first) return;
  await exec(
    "INSERT INTO panel_users (username, password_hash, two_fa_method, is_active, created_at, updated_at) VALUES (?, ?, 'none', 1, ?, ?)",
    [config.panelAdminUsername, hashPassword(config.panelAdminPassword), new Date().toISOString(), new Date().toISOString()]
  );
}

export async function cleanupExpiredAuth() {
  const db = getDb();
  if (db.dialect === "mysql") {
    await exec("DELETE FROM panel_sessions WHERE expires_at < NOW()", []);
    await exec("DELETE FROM two_fa_challenges WHERE expires_at < NOW()", []);
    return;
  }
  const now = nowIso();
  await exec("DELETE FROM panel_sessions WHERE expires_at < ?", [now]);
  await exec("DELETE FROM two_fa_challenges WHERE expires_at < ?", [now]);
}

export async function createPanelSession(userId, ip = "", userAgent = "") {
  await cleanupExpiredAuth();
  const token = randomBytes(32).toString("base64url");
  const db = getDb();
  if (db.dialect === "mysql") {
    await exec(
      "INSERT INTO panel_sessions (token, user_id, expires_at, created_at, ip_address, user_agent) VALUES (?, ?, DATE_ADD(NOW(), INTERVAL ? HOUR), NOW(), ?, ?)",
      [token, userId, config.panelSessionHours, String(ip || "").slice(0, 45), String(userAgent || "").slice(0, 512)]
    );
    const row = await queryOne("SELECT expires_at FROM panel_sessions WHERE token = ?", [token]);
    return { token, expires: row?.expires_at };
  }
  const expires = isoAfterHours(config.panelSessionHours);
  await exec(
    "INSERT INTO panel_sessions (token, user_id, expires_at, created_at, ip_address, user_agent) VALUES (?, ?, ?, ?, ?, ?)",
    [token, userId, expires, nowIso(), String(ip || "").slice(0, 45), String(userAgent || "").slice(0, 512)]
  );
  return { token, expires };
}

export async function getPanelSession(token) {
  if (!token) return null;
  const db = getDb();
  const row = db.dialect === "mysql"
    ? await queryOne(
      "SELECT u.* FROM panel_sessions s JOIN panel_users u ON u.id = s.user_id WHERE s.token = ? AND s.expires_at >= NOW() AND u.is_active = 1",
      [token]
    )
    : await queryOne(
      "SELECT u.* FROM panel_sessions s JOIN panel_users u ON u.id = s.user_id WHERE s.token = ? AND s.expires_at >= ? AND u.is_active = 1",
      [token, new Date().toISOString()]
    );
  return row || null;
}

export async function revokePanelSession(token) {
  if (!token) return;
  await exec("DELETE FROM panel_sessions WHERE token = ?", [token]);
}

export async function listUserSessions(userId) {
  await cleanupExpiredAuth();
  const db = getDb();
  if (db.dialect === "mysql") {
    return queryAll("SELECT * FROM panel_sessions WHERE user_id = ? AND expires_at >= NOW() ORDER BY created_at DESC", [userId]);
  }
  return queryAll("SELECT * FROM panel_sessions WHERE user_id = ? AND expires_at >= ? ORDER BY created_at DESC", [userId, new Date().toISOString()]);
}

export async function revokeSessionToken(userId, token) {
  const r = await exec("DELETE FROM panel_sessions WHERE user_id = ? AND token = ?", [userId, token]);
  return r.changes > 0;
}

export async function revokeOtherSessions(userId, keepToken) {
  const r = keepToken
    ? await exec("DELETE FROM panel_sessions WHERE user_id = ? AND token <> ?", [userId, keepToken])
    : await exec("DELETE FROM panel_sessions WHERE user_id = ?", [userId]);
  return r.changes;
}

export async function getPrimaryPanelUser() {
  return queryOne("SELECT * FROM panel_users WHERE is_active = 1 ORDER BY id LIMIT 1", []);
}

export async function create2faChallenge(user, method, purpose = "login", code = "") {
  await cleanupExpiredAuth();
  const token = randomBytes(24).toString("base64url");
  const finalCode = method === "telegram" && !code ? String(Math.floor(100000 + Math.random() * 900000)) : String(code || "");
  const db = getDb();
  if (db.dialect === "mysql") {
    await exec(
      "INSERT INTO two_fa_challenges (token, user_id, code, method, purpose, expires_at, created_at) VALUES (?, ?, ?, ?, ?, DATE_ADD(NOW(), INTERVAL ? MINUTE), NOW())",
      [token, user.id, finalCode, method, purpose, CHALLENGE_MINUTES]
    );
    const row = await queryOne("SELECT expires_at FROM two_fa_challenges WHERE token = ?", [token]);
    return { token, code: finalCode, expires: row?.expires_at };
  }
  const expires = isoAfterMinutes(CHALLENGE_MINUTES);
  await exec(
    "INSERT INTO two_fa_challenges (token, user_id, code, method, purpose, expires_at, created_at) VALUES (?, ?, ?, ?, ?, ?, ?)",
    [token, user.id, finalCode, method, purpose, expires, nowIso()]
  );
  return { token, code: finalCode, expires };
}

export async function find2faChallenge(challengeToken, userId, purpose = null) {
  const db = getDb();
  const params = [challengeToken, userId];
  let sql = "SELECT * FROM two_fa_challenges WHERE token = ? AND user_id = ?";
  if (purpose) {
    sql += " AND purpose = ?";
    params.push(purpose);
  }
  if (db.dialect === "mysql") {
    sql += " AND expires_at >= NOW()";
  } else {
    sql += " AND expires_at >= ?";
    params.push(new Date().toISOString());
  }
  return queryOne(sql, params);
}

export async function verify2faChallenge(challengeToken, user, code, method = null, purpose = null) {
  const row = await find2faChallenge(challengeToken, user.id, purpose);
  if (!row) return false;
  const useMethod = method || row.method;
  if (row.method !== useMethod) return false;
  const entered = String(code || "").trim().replaceAll(" ", "");
  let ok = false;
  if (useMethod === "telegram") ok = row.code === entered;
  if (useMethod === "totp" && user.totp_secret) ok = authenticator.check(entered, user.totp_secret);
  if (ok) await exec("DELETE FROM two_fa_challenges WHERE token = ?", [challengeToken]);
  return ok;
}

export function generateTotpSetup(user) {
  const secret = authenticator.generateSecret();
  const uri = authenticator.keyuri(user.username, config.panelTotpIssuer, secret);
  return { secret, uri };
}

export function verifyTotpCode(user, code) {
  if (!user?.totp_secret) return false;
  return authenticator.check(String(code || "").trim().replaceAll(" ", ""), user.totp_secret);
}

export function loginRequires2fa(user) {
  return ["totp", "telegram"].includes((user.two_fa_method || "none").toLowerCase());
}

export function sessionResponse(user, token, expires) {
  return { access_token: token, expires_at: expires, user: { id: user.id, username: user.username }, auth_type: "panel" };
}
